src/message/mailbox/parsers.rs

```diff
@@ -0,0 +1,322 @@
+use chumsky::prelude::*;
+
+// Core Rules
+// https://datatracker.ietf.org/doc/html/rfc2234#section-6.1
+
+// CRLF           =  CR LF
+//                        ; Internet standard newline
+pub fn crlf() -> impl Parser<char, Vec<char>, Error = Simple<char>> {
+    just('\r').chain(just('\n'))
+}
+
+// WSP            =  SP / HTAB
+//                        ; white space
+pub fn wsp() -> impl Parser<char, char, Error = Simple<char>> {
+    one_of([' ', '\t'])
+}
+
+// Primitive Tokens
+// https://datatracker.ietf.org/doc/html/rfc2822#section-3.2.1
+
+// NO-WS-CTL       =       %d1-8 /         ; US-ASCII control characters
+//                         %d11 /          ;  that do not include the
+//                         %d12 /          ;  carriage return, line feed,
+//                         %d14-31 /       ;  and white space characters
+//                         %d127
+pub fn no_ws_ctl() -> impl Parser<char, char, Error = Simple<char>> {
+    let mut range: Vec<char> = vec![];
```

Comment:
It seems like this code is instantiating this function many times every time it uses a parser. That will include allocating dozens of these vectors. Am I missing some magic or is this making parsing very expensive.

Is there a way we can instantiate this parser just once and reuse it? Or ideally reference the same data for each branch rather than having many unique character sets that are all identical.

Or is there some magic/optimization that keeps the memory usage down and dedups the code?

Reply:
Would `once_cell::sync::Lazy` fix the problem?

Reply:
```rust

static TEXT: Lazy<Vec<char>> = Lazy::new(|| {
    let mut range: Vec<char> = vec![];
    range.extend((1 as char)..=(9 as char));
    range.push(11 as char);
    range.push(12 as char);
    range.extend((14 as char)..=(127 as char));
    range
});

fn text() -> impl Parser<char, char, Error = Simple<char>> {
    one_of(TEXT.as_slice())
}
```

Reply:
Yeah, this seems good enough. I'm surprised the chumsky doesn't have a recommendation here as it seems like a common issue.